### ISSUE TITLE
Where is the Dockerfile of the pgbouncer image?

### ISSUE THREAD

**Issue by pierreozoux** [zalando/postgres-operator#1964]
I plan to use a docker image to have pgbouncer in a cluster, but for something else than postgres-operator, and I was wondering where is the code of the image so that I can reuse it.
(Seems like the other option is bitnami or "personal" image, I'd prefer to rely on zalando and its beautiful community!)

Thanks :)

**Comment by FxKu** [comment]
We haven't yet open sourced the image. It also includes some patches to pgBouncer. Would like to publish, but other topics had a higher priority so far

**Comment by szelenka** [comment]
I'm running into a similar situation, where it seems there's some custom config applied to the pgbouncer image that expects these lines in the `pgbouncer.ini.tmpl` file:
```
stats_users_prefix = robot_
```

When the open source pgbouncer attempts to start with that in the ini file, it will error out. Any insight into what Spilo is doing with pgbouncer to expect that in the ini file would be welcome!

**Comment by FactorT** [comment]
+1 for publishing Dockerfile of pgBouncer

**Comment by francoispqt** [comment]
+1 for publishing the Dockerfile, or at least be transparent on what are the patches to pgbouncer. We want to know what we are running on our system.

**Comment by andyndang** [comment]
+1 for publishing the docker image. We want to deploy to ARM64 and this is currently blocking us

**Comment by bwrobc** [comment]
+1 for publishing the Dockerfile and PGBouncer patches, for all the above reasons. It's been about nine months since this issue was opened.

**Comment by raviranjanelastisys** [comment]
Hi Guys

Any luck on this ? Will really appreciate if you can move it forward.

**Comment by kaiba42** [comment]
+1 for Dockerfile and/or patches. Trying to deploy to a cluster running both arm64 and amd64 nodes, and can't set a nodeSelector for the connectionPooler settings in Postgresql custom resource yaml.

**Comment by tarunmittal24** [comment]
+1, please upload the dockerfile, needs to update the pgbouncer version and other settings like ssl

**Comment by raviranjanelastisys** [comment]
Any update here?

**Comment by dwyanepolinski** [comment]
+1, pgbouncer 1.21 was released with added supprot for prepared statements https://www.pgbouncer.org/2023/10/pgbouncer-1-21-0. The newest version of pgbouncer image is 1.20 and its even unsafe to create own build

**Comment by wzrdtales** [comment]
This can be reconstructed through the history


```
ENTRYPOINT ["/bin/sh" "/entrypoint.sh"]USER pgbouncer:pgbouncerRUN chown -R pgbouncer:pgbouncer /var/log/pgbouncer /var/run/pgbouncer /etc/pgbouncer /etc/ssl/certsADD file:b786bad4b80758c015464593955e6c0bfc6909746c39c0ded33335db3ac5ecaf in ./ADD file:4f6edf15149a40369c216c4da6706e69f622a1432739e77b3dbaf69d115d403c in /etc/pgbouncer/ADD file:80e4d3a584326304f10ea949bd4641f3e52a5693c7b950fe7afb40518dfd3475 in /etc/pgbouncer/COPY file:45132f8b9fc583334b10ac78d68178c82d5d0a2a37d1c08563b92a67be17ffac in /bin/pgbouncerRUN addgroup -S pgbouncer && adduser -S pgbouncer && mkdir -p /etc/pgbouncer /var/log/pgbouncer /var/run/pgbouncerWORKDIR /RUN apk --update add libevent openssl c-ares gettext ca-certificates postgresql-clientCOPY zalando-marker / # buildkitRUN apk update && apk upgrade && rm -rf /var/cache/apk/* /tmp/* # buildkitCMD ["/bin/sh"]ADD file:37a76ec18f9887751cd8473744917d08b7431fc4085097bb6a09d81b41775473 in /

ENTRYPOINT ["/bin/sh" "/entrypoint.sh"]
USER pgbouncer:pgbouncer

RUN chown -R pgbouncer:pgbouncer /var/log/pgbouncer /var/run/pgbouncer /etc/pgbouncer /etc/ssl/certs

ADD file:b786bad4b80758c015464593955e6c0bfc6909746c39c0ded33335db3ac5ecaf in ./
ADD file:4f6edf15149a40369c216c4da6706e69f622a1432739e77b3dbaf69d115d403c in /etc/pgbouncer/
ADD file:80e4d3a584326304f10ea949bd4641f3e52a5693c7b950fe7afb40518dfd3475 in /etc/pgbouncer/
COPY file:45132f8b9fc583334b10ac78d68178c82d5d0a2a37d1c08563b92a67be17ffac in /bin/pgbouncer
RUN addgroup -S pgbouncer && adduser -S pgbouncer && mkdir -p /etc/pgbouncer /var/log/pgbouncer /var/run/pgbouncer
WORKDIR /
RUN apk --update add libevent openssl c-ares gettext ca-certificates postgresql-client
COPY zalando-marker / # buildkit
RUN apk update && apk upgrade && rm -rf /var/cache/apk/* /tmp/* # buildkit
CMD ["/bin/sh"]
ADD file:37a76ec18f9887751cd8473744917d08b7431fc4085097bb6a09d81b41775473 in /
```

So the missing part is actually the compiled pgbouncer binary. So given that, building an image from the latest pgbouncer image should be fine if it is just about adding things to the config.

**Comment by L1ghtman2k** [comment]
What is the usecase for the robot user? If we put our own pgbouncer image, what would we need to expect to have on it?

**Comment by mamccorm** [comment]
Hey @FxKu, any updates here? we're hoping to leverage this operator in a secure environment but have concerns with the amount of vulnerabilities, as well as there being closed-source modifications to pgbouncer. It also looks like there is potentially another zalando-marker application at play as well.

Can you publish the source for those? I'm guessing they are light patches to make compatible with the operators settings. or alternatively, publish them as archives somewhere that users could re-use to build their own image?

**Comment by slimm609** [comment]
agreed.  This is a major issue of zalando doing closed-source modifications in what is already an open source product and releases it publicly.

**Comment by marcinpolinski** [comment]
Found the solution for this, just don't use postgres-operator from zalando (I switched to cloudnative-pg, way better)

**Comment by Sitronik** [comment]
Not long ago I also encountered this problem. I needed to use pgbouncer no lower than version 1.23.0 so that I could use the server_lifetime parameter. I made an image with the latest version 1.24.0. I also disabled the stats_users_prefix parameter without it, everything works fine, I didn’t find any problems, judging by the name of the parameter most likely it relates to some kind of statistics

```dockerfile
FROM --platform=linux/amd64 registry.opensource.zalan.do/acid/pgbouncer:master-32 as build

USER root

RUN apk update && \
    apk add git build-base libtool m4 automake autoconf pandoc-cli pkgconf libevent-dev openssl-dev

RUN git clone https://github.com/pgbouncer/pgbouncer.git
RUN cd pgbouncer && git submodule init && git submodule update && ./autogen.sh && ./configure && make && make install

RUN cp /usr/local/bin/pgbouncer /bin/pgbouncer

FROM --platform=linux/amd64 registry.opensource.zalan.do/acid/pgbouncer:master-32 as image

#RUN pgbouncer --version

USER root

COPY --from=build /usr/local/bin/pgbouncer /bin/pgbouncer

RUN apk add --upgrade apk-tools &&  \
    apk update && apk -U upgrade && \
    rm -rf /var/cache/apk/*

USER pgbouncer

RUN sed -i '16s/stats_users_prefix = robot_/#stats_users_prefix = robot_/' /etc/pgbouncer/pgbouncer.ini.tmpl
```

**Comment by yuvalavidor** [comment]
+1 for this, the minimum can be adding arm support

**Comment by kuzaxak** [comment]
Waiting to see arm support too

**Comment by FxKu** [comment]
There it is: https://github.com/zalando/postgres-operator/blob/master/pooler/Dockerfile

The multi-arch image is published under: ghcr.io/zalando/postgres-operator/pgbouncer:v2.0.0
